src/main/java/org/opensearch/ad/feature/CompositeRetriever.java

```diff
@@ -0,0 +1,346 @@
+        this.pageSize = pageSize;
+        this.expirationEpochMs = expirationEpochMs;
+        this.clock = clock;
+    }
+
+    // a constructor that provide default value of clock
+    public CompositeRetriever(
+        long dataStartEpoch,
+        long dataEndEpoch,
+        AnomalyDetector anomalyDetector,
+        NamedXContentRegistry xContent,
+        Client client,
+        long expirationEpochMs,
+        Settings settings,
+        int maxEntitiesPerInterval,
+        int pageSize
+    ) {
+        this(
+            dataStartEpoch,
+            dataEndEpoch,
+            anomalyDetector,
+            xContent,
+            client,
+            expirationEpochMs,
+            Clock.systemUTC(),
+            settings,
+            maxEntitiesPerInterval,
+            pageSize
+        );
+    }
+
+    public void start(ActionListener<Page> listener) {
+        try {
+            RangeQueryBuilder rangeQuery = new RangeQueryBuilder(anomalyDetector.getTimeField())
+                .gte(dataStartEpoch)
+                .lt(dataEndEpoch)
+                .format("epoch_millis");
+
+            BoolQueryBuilder internalFilterQuery = new BoolQueryBuilder().filter(anomalyDetector.getFilterQuery()).filter(rangeQuery);
+
+            // multiple categorical fields are supported
+            CompositeAggregationBuilder composite = AggregationBuilders
+                .composite(
+                    AGG_NAME_COMP,
+                    anomalyDetector
+                        .getCategoryField()
+                        .stream()
+                        .map(f -> new TermsValuesSourceBuilder(f).field(f))
+                        .collect(Collectors.toList())
+                )
+                .size(pageSize);
+            for (Feature feature : anomalyDetector.getFeatureAttributes()) {
+                AggregatorFactories.Builder internalAgg = ParseUtils
+                    .parseAggregators(feature.getAggregation().toString(), xContent, feature.getId());
+                composite.subAggregation(internalAgg.getAggregatorFactories().iterator().next());
+            }
+
+            // In order to optimize the early termination it is advised to set track_total_hits in the request to false.
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+                .query(internalFilterQuery)
+                .size(0)
+                .aggregation(composite)
+                .trackTotalHits(false);
+
+            Page page = new Page(null, searchSourceBuilder, null);
+            page.next(client, anomalyDetector, listener);
+
+        } catch (Exception e) {
+            listener
+                .onFailure(new EndRunException(anomalyDetector.getDetectorId(), CommonErrorMessages.INVALID_SEARCH_QUERY_MSG, e, false));
+        }
+    }
+
+    public class Page {
+        // a map from categorical field name to values (type: java.lang.Comparable)
+        Map<String, Object> afterKey;
```

Comment:
Does the listener need to know the afterKey and source? IMO, the listener should only consume the results.

Reply:
no, the listener does not need to know the afterkey and source.